libs/zedUpload/httputil/http.go

```diff
@@ -148,30 +150,112 @@ func ExecCmd(cmd, host, remoteFile, localFile string, objSize int64,
 			return stats
 		}
 		defer local.Close()
-		defer resp.Body.Close()
-		chunkSize := SingleMB
-		var written, copiedSize int64
-		stats.Size = objSize
-		for {
-			var copyErr error
-			if written, copyErr = io.CopyN(local, resp.Body, chunkSize); copyErr != nil && copyErr != io.EOF {
-				stats.Error = copyErr
-				return stats
+
+		var errorList []string
+		done := false
+		canSeek := false
+
+		delay := time.Second
+		lastModified := ""
+		var currentErrorList []string
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			if len(currentErrorList) > 0 {
+				// if we have previous errors, collect them
+				errorList = append(errorList, fmt.Sprintf("(attempt %d/%d):", attempt, maxRetries))
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				currentErrorList = []string{}
 			}
-			copiedSize += written
-			if written != chunkSize {
-				// Must have reached EOF
+			if ctx.Err() != nil {
+				errorList = append(errorList, fmt.Sprintf("ctx.Err: %v", ctx.Err()))
 				break
 			}
-			stats.Asize = copiedSize
-			if prgNotify != nil {
-				select {
-				case prgNotify <- stats:
-				default: //ignore we cannot write
+			if done {
+				break
+			}
```

Comment:
Why repeat the error checks up front and check done up front and not at the end of the loop?
Also see below; the setting of done seems to be incomplete.

Reply:
Actually it is here for simplify continue commands logic below, but good point, will think how to refactor it.

Reply:
But you didn't have any continue statements in the loop afair

Reply:
Actually in case of errors from CopyN (not EOF) we will go to the upper loop with attempts.